crossbeam-skiplist/src/base.rs

```diff
@@ -2121,3 +2122,252 @@ fn below_upper_bound<T: Ord + ?Sized>(bound: &Bound<&T>, other: &T) -> bool {
         Bound::Excluded(key) => other < key,
     }
 }
+
+/// A reference-counted entry in a skip list.
+pub struct RefCountedEntry<K, V> {
+    node: *const Node<K, V>,
+}
+
+unsafe impl<K, V> Send for RefCountedEntry<K, V> {}
+
+impl<K, V> Drop for RefCountedEntry<K, V> {
+    fn drop(&mut self) {
+        let guard = &epoch::pin();
+        unsafe { (*self.node).decrement(guard) }
+    }
+}
+
+impl<K, V> RefCountedEntry<K, V> {
+    /// Tries to create a new `RefCountedEntry` by incrementing the reference count of
+    /// a node.
+    fn try_acquire(node: &Node<K, V>) -> Option<RefCountedEntry<K, V>> {
+        if unsafe { node.try_increment() } {
+            Some(RefCountedEntry {
+                node: node as *const _,
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Returns a reference to the key.
+    pub fn key(&self) -> &K {
+        unsafe { &(*self.node).key }
+    }
+
+    /// Returns a reference to the value.
+    pub fn value(&self) -> &V {
+        unsafe { &(*self.node).value }
+    }
+}
+
+/// A lock-free skip list.
+pub struct ConcurrentSkipList<K, V> {
+    /// The underlying skip list
+    list: Arc<SkipList<K, V>>,
```

Comment:
I think we could add a method that takes `self: Arc<Self>` instead of creating a new type. Like:

- https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#method.lock_owned
- https://docs.rs/async-lock/latest/async_lock/struct.Mutex.html#method.lock_arc

Reply:
Thanks for the reply. It's a bit late here, I will see it tomorrow.

Reply:
I have modified the code. PTAL again, thanks @taiki-e . If this is desired, I will add more tests.